src/lib/idl.ml

```diff
@@ -118,14 +118,15 @@ let get_arg call has_named name is_opt =
     match unnamed with
     | head :: tail ->
         Result.Ok (head, {call with Rpc.params= Rpc.Dict named :: tail})
-    | _ -> Result.Error (`Msg "Incorrect number of arguments") )
+    | _ -> Result.Error (`Msg "Incorrect number of argumentss") )
   | true, _, _ ->
       Result.Error
         (`Msg
           "Marshalling error: Expecting dict as first argument when named \
            parameters exist")
   | false, None, head :: tail -> Result.Ok (head, {call with Rpc.params= tail})
-  | false, None, [] -> Result.Error (`Msg "Incorrect number of arguments")
+  | false, None, [] -> Result.Ok (Null, call)
+    (* don't understand this one : Result.Error (`Msg "Incorrect number of arguments") *)
```

Comment:
I don't understand the former behaviour here. Why does it fail on my request? If you're interested, the request comes from emacs with lsp-mode 6.1... I think the request should be accepted but it failed with the old behaviour nevertheless. I will remove the comment anyway...

Reply:
IIRC it hits this clause when it's looking for the next unnamed argument, but there aren't any left.

Reply:
Hmmmm, ok. But why does it try to retrieve an unnamed argument in my case?

Reply:
It's called up in the body of `declare` in the `GenServer` modules [here](https://github.com/mirage/ocaml-rpc/blob/6f81568a86d33cb3e8fa62dbe6bc3f3ac3fb052c/src/lib/idl.ml#L371). The implication is that you've declared the call to have an argument that it's trying to extract from the `call` structure, but there's nothing left.

Reply:
How can I declare a call with no arguments at all? I used `()` to model that behaviour but it seems it's not working... Should I just use `returning t e` ? But then I would have to return a constant, which is not necessarily the desired behaviour.

Reply:
no, that's not returning a constant, it's returning a constant _type_

Reply:
I might not be sure to get my point across: if I define a function, like, `let init = declare_notification "initialized" [] (returning unit_p err)` which doesn't take any argument, then I have to associate it with the actual implementation, like `Server.init (MyIdl.ErrM.return ());` or something like that. The argument to Server.init will be of type `(unit, error) result M.T.box`, which is a constant and not a function, since it doesn't take any arguments (not even `()).` But I'd like to have a behaviour depending on state, even if I'm not using any arguments from the JSON-RPC call.

Reply:
Ah, I see what you mean. You could try rebinding it when you're processing the request? e.g., modifying example2:

```ocaml
let start_server () =
  Server.query query;
  Server.diagnostics diagnostics;
  Server.test test;

  let rpc_fn = MyIdl.server Server.implementation in

  let process x =
    let open M in
    Server.query query; (* <--- rebinding to a new function here *)
    rpc_fn (Jsonrpc.call_of_string (Bytes.unsafe_to_string x))
    >>= fun response -> Jsonrpc.string_of_response response |> return in

  serve_requests process sockpath
```

It's a bit ugly though.

Reply:
I think it would work, but as you say, it's not as beautiful as it could be. Just to know, what are the semantics associated to the unit type, in ocaml-rpc?

Reply:
Ok, how about this. We could add a special case for when an RPC takes only optional named arguments. Currently an invocation of this with all arguments 'None' would be marshalled as an empty record as the first argument, but we could notice this situation and handle there being no arguments at all. Then you could define your function as taking one optional argument that your implementation could then discard. Again pretty ugly, but the only alternatives I can think of are things like requiring each implementation function to take a unit as last / first argument or something similar.

Reply:
It could work that way but I think before engineering a solution on ocaml-rpc's side, I'll use something inspired by the snippet you gave me before, and think about it to see if I see something better to handle my usecase. Are you ok with that?

Reply:
Of course!